src/fsharp/FSharp.Core/quotations.fs

```diff
@@ -91,14 +91,16 @@ open Helpers
 [<CompiledName("FSharpVar")>]
 [<System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage","CA2218:OverrideGetHashCodeOnOverridingEquals",Justification="Equals override does not equate further objects, so default GetHashCode is still valid")>]
 type Var(name: string, typ:Type, ?isMutable: bool) =
-
     inherit obj()
-    static let mutable lastStamp = 0L
+
+    static let getStamp =
+        let mutable lastStamp = -1L // first value retrieved will be 0
+        fun () -> System.Threading.Interlocked.Increment &lastStamp
```

Comment:
Lock would have been fine if the value being assigned was from within the lock, but it wasn't. But yes, interlocked is preferable as well.